### ISSUE TITLE
Easiest way to disable rounded corners (no patching/adapter required)

### ISSUE THREAD

**Issue by limacohen** [valinet/Win11DisableRoundedCorners#42]
I stumbled across this preference in [PowerToys](https://github.com/microsoft/PowerToys). If you just want to disable rounded corners and you don't mind using FancyZones, this is by far the easiest (official?) way to do it:
!["Disable round corners when window is snapped"](https://user-images.githubusercontent.com/541169/204914289-d64b4396-4557-4baf-a0df-4e2c6b284e79.png)

**Comment by rich-ayr** [comment]
Interesting, looks like it's possible to set this on a per-window basis using [DwmSetWindowAttribute](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmsetwindowattribute) + [DWM_WINDOW_CORNER_PREFERENCE](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwm_window_corner_preference).

I wonder if this could get rid of the need for running with elevation. One minor issue I see with this method is the need to install a window hook or have a thread polling to account for new windows. This is also only (officially) supported starting with build 22000.

**Comment by valinet** [comment]
> Interesting, looks like it's possible to set this on a per-window basis using [DwmSetWindowAttribute](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmsetwindowattribute) + [DWM_WINDOW_CORNER_PREFERENCE](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwm_window_corner_preference).
> 
> I wonder if this could get rid of the need for running with elevation. One minor issue I see with this method is the need to install a window hook or have a thread polling to account for new windows.

You cannot get around elevation. As a principle, an app cannot “alter” another app which is on a higher privilege level.

ExplorerPatcher already uses this API for Simple Windows Switcher’s “corner” preference.

The thing with this API is that making use of it reliably is more cumbersome:
1. To modify elevated apps, you need to run elevated.
2. You need a global hook on window creation, something like a shell hook and looking for the HSHELL_RUDEAPPACTIVATED or HSHELL_WINDOWACTIVATED. Maybe additionally a “winevent” hook… the strategy used by sws can be used and can detect new windows that are created and alter them.
3. Of course, apps can still override what you do, this largely being a hack.
4. There may be trouble altering protected processes or protected processes light, and no real solution might be there other than creating a service…
5. It might be more reliable to continuously poll the window list and set the property on all windows, but or course this consumes tons of resources for nothing really. Even with a hook, it consumes more compared to previous, well established methods.
6. With the method employed by ep_dwm or win11-toggle-rounded-corners, for example, it’s just is more reliable and always respected: the compositor is forced to always draw right angle corners, so all windows are affected.

So yeah, idk, I wouldn’t call this easy at all. The method employed by ep_dwm (ExplorerPatcher) is the technically correct strategy. Rounded corners appeared only since around build 22000, so I don’t see the problem here - this API works basically since Windows 11 RTM, like the Mica backdrop brush and so on.

**Comment by limacohen** [comment]
> So yeah, idk, I wouldn’t call this easy at all.

I mean, it's easy from the user point of view. I already use FancyZones so I just had to click a checkbox. I'm sure this would be the case for other people too. I assume the window hook was already in place for moving/resizing the windows.

Just to be clear, I enormously appreciate the effort done by both of you guys. I'm just shedding some light into an alternative way to get to the same(ish) thing.

**Comment by valinet** [comment]
> > So yeah, idk, I wouldn’t call this easy at all.
> 
> I mean, it's easy from the user point of view. I already use FancyZones so I just had to click a checkbox. I'm sure this would be the case for other people too. I assume the window hook was already in place for moving/resizing the windows.
> 
> Just to be clear, I enormously appreciate the effort done by both of you guys. I'm just shedding some light into an alternative way to get to the same(ish) thing.

Yeah, idk, as I said, no doubt it can work. Everybody can decide what works best for them. I looked on FancyZones: the fact that it is a first party product yet it employs so many hacks against their own operating system says a lot about how friendly developing on Windows has become.

As I said, it probably runs some elevated agent in order to apply the property to elevated windows. Do PowerToys require admin when installing…?

**Comment by rich-ayr** [comment]
I agree that this method does require more work to get working perfectly and it's not really worth the effort.

> You cannot get around elevation. As a principle, an app cannot “alter” another app which is on a higher privilege level.

Well you can get a window handle to every process without elevation, that's what motivated my thinking here.
Taking a quick glance at DwmSetWindowAttribute there is some access checks happening in win32k, cba to look into it any further though.

> As I said, it probably runs some elevated agent in order to apply the property to elevated windows. Do PowerToys require admin when installing…?

The installer seems to [schedule](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs#L331) an [elevated](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs#L151) [task](https://github.com/microsoft/PowerToys/blob/db191b8b75ad456d5eb94eaa0aa0c7b59ecec900/installer/PowerToysSetupCustomActions/CustomAction.cpp#L245).
Being MSFT they could even digitally sign using Windows Publisher and casually bypass UAC prompts ([autoElevate](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src/SbsCs/application-manifests.md#autoElevate)) if they chose to.
I think it's quite sad that just wanting to install an app to be run as admin is such a hassle on this platform.
